blackboxopt/optimizers/botorch_base.py

```diff
@@ -0,0 +1,463 @@
+    ):
+        """Single objective BO optimizer that uses as a surrogate model the `model`
+        object provided by user.
+
+        The `model` is expected to be extended from BoTorch base model `Model` class,
+        and does not require to be a GP model.
+
+        Args:
+            search_space: The space in which to optimize.
+            objective: The objective to optimize.
+            model: Surrogate model of `Model` type.
+            acquisition_function_factory: Callable that produces an acquisition function
+                instance, could also be a compatible acquisition function class.
+                Only acquisition functions to be minimized are supported.
+                Providing a partially initialized class is possible with, e.g.
+                `functools.partial(UpperConfidenceBound, beta=6.0, maximize=False)`.
+            af_optimizer_kwargs: Settings for acquisition function optimizer,
+                see `botorch.optim.optimize_acqf`.
+            num_initial_random_samples: Size of the initial space-filling design that
+                is used before starting BO. The points are sampled randomly in the
+                search space. If no random sampling is required, set it to 0.
+            max_pending_evaluations: Maximum number of parallel evaluations. For
+                sequential BO use the default value of 1. If no limit is required,
+                set it to None.
+            batch_shape: Batch dimension(s) used for batched models.
+            logger: Custom logger.
+            seed: A seed to make the optimization reproducible.
+            torch_dtype: Torch data type used for storing the data. This needs to match
+                the dtype of the model used
+        """
+        super().__init__(search_space=search_space, objective=objective, seed=seed)
+        self.num_initial_random = num_initial_random_samples
+        self.max_pending_evaluations = max_pending_evaluations
+        self.batch_shape = batch_shape
+        self.logger = logger or logging.getLogger("blackboxopt")
+
+        self.torch_dtype = torch_dtype
+        self.X = torch.empty(
+            (*self.batch_shape, 0, len(search_space)), dtype=torch_dtype
+        )
+        self.losses = torch.empty((*self.batch_shape, 0, 1), dtype=torch_dtype)
+        self.pending_specifications: Dict[int, EvaluationSpecification] = {}
+        if seed is not None:
+            torch.manual_seed(seed=seed)
+
+        self.model = model
+        self.acquisition_function_factory = acquisition_function_factory
+        self.af_opt_kwargs = init_af_opt_kwargs(af_optimizer_kwargs)
+
+    def _create_fantasy_model(self, model: Model) -> Model:
+        """Create model with the pending specifications and model based
+        outcomes added to the training data."""
+        if not self.pending_specifications:
+            # nothing to do when there are no pending specs
+            return model
+
+        pending_X = torch.tensor(
```

Comment:
`pending_X` needs to be given `dtype=self.torch_dtype`, otherwise this will always end up being a `float64` tensor, as this is the numpy default.